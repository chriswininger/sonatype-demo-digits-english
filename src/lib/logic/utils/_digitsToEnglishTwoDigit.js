/*
    Chris Wininger
    _digitsToEnglishTwoDigit

    An internal helper function, which special cases the problem of converting numbers to english text
        representations to two digits. This should not be exposed outside the API except to test
 */

export default (strNum) => {
    if (typeof strNum !== 'string')
        throw new TypeError('_digitsToEnglishTwoDigit only accepts strings')

    strNum = strNum.replace(/^0+/, '')

    if (strNum.length > 2)
        throw new RangeError('_digitsToEnglishTwoDigit accepts at most two characters')

    if (strNum.length === 0)
        return ''

    const wordMapSingle = {
        '0': 'zero',
        '1': 'one',
        '2': 'two',
        '3': 'three',
        '4': 'four',
        '5': 'five',
        '6': 'six',
        '7': 'seven',
        '8': 'eight',
        '9': 'nine',
        '10': 'ten',
        '11': 'eleven',
        '12': 'twelve',
        '13': 'thirteen',
        '14': 'fourteen',
        '15': 'fifteen',
        '16': 'sixteen',
        '17': 'seventeen',
        '18': 'eighteen',
        '19': 'nineteen',
    }

    const wordMapTens = {
        '2': 'twenty',
        '3': 'thirty',
        '4': 'forty',
        '5': 'fifty',
        '6': 'sixty',
        '7': 'seventy',
        '8': 'eighty',
        '9': 'ninety'
    }

    if (wordMapSingle[strNum])
        return wordMapSingle[strNum]

    if (strNum[1] !== '0')
        return wordMapTens[strNum[0]] + ' ' + wordMapSingle[strNum[1]]
    else
        return wordMapTens[strNum[0]]
}