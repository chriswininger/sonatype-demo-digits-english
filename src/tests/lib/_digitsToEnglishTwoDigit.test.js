import _digitsToEnglishTwoDigit from '../../lib/logic/utils/_digitsToEnglishTwoDigit'

// === Check Validation ===
it('throws TypeError when a non-string value is passed', () => {
    expect(() => _digitsToEnglishTwoDigit({ value: 'some object' })).toThrowErrorMatchingSnapshot()
})

it('throws a RangeError when a value with more than two digits is used', () => {
    expect(() => _digitsToEnglishTwoDigit('100')).toThrowErrorMatchingSnapshot()
    expect(() => _digitsToEnglishTwoDigit('0100')).toThrowErrorMatchingSnapshot()
})

// === Check Input cleanup ===
it('evaluates numbers with leading zeros', () => {
    expect(_digitsToEnglishTwoDigit('001')).toBe('one')
    expect(_digitsToEnglishTwoDigit('0012')).toBe('twelve')
})

// === Alert Us To Any Changes from Here on out to valid input ===
it('numbers match expected output (full range test)', () => {
    for (let i = 0; i < 100; i++) {
        expect(_digitsToEnglishTwoDigit('' + i)).toMatchSnapshot()
    }
})