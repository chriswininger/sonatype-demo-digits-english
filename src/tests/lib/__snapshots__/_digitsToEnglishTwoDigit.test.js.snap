// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`numbers match expected output (full range test) 1`] = `""`;

exports[`numbers match expected output (full range test) 2`] = `"one"`;

exports[`numbers match expected output (full range test) 3`] = `"two"`;

exports[`numbers match expected output (full range test) 4`] = `"three"`;

exports[`numbers match expected output (full range test) 5`] = `"four"`;

exports[`numbers match expected output (full range test) 6`] = `"five"`;

exports[`numbers match expected output (full range test) 7`] = `"six"`;

exports[`numbers match expected output (full range test) 8`] = `"seven"`;

exports[`numbers match expected output (full range test) 9`] = `"eight"`;

exports[`numbers match expected output (full range test) 10`] = `"nine"`;

exports[`numbers match expected output (full range test) 11`] = `"ten"`;

exports[`numbers match expected output (full range test) 12`] = `"eleven"`;

exports[`numbers match expected output (full range test) 13`] = `"twelve"`;

exports[`numbers match expected output (full range test) 14`] = `"thirteen"`;

exports[`numbers match expected output (full range test) 15`] = `"fourteen"`;

exports[`numbers match expected output (full range test) 16`] = `"fifteen"`;

exports[`numbers match expected output (full range test) 17`] = `"sixteen"`;

exports[`numbers match expected output (full range test) 18`] = `"seventeen"`;

exports[`numbers match expected output (full range test) 19`] = `"eighteen"`;

exports[`numbers match expected output (full range test) 20`] = `"nineteen"`;

exports[`numbers match expected output (full range test) 21`] = `"twenty"`;

exports[`numbers match expected output (full range test) 22`] = `"twenty-one"`;

exports[`numbers match expected output (full range test) 23`] = `"twenty-two"`;

exports[`numbers match expected output (full range test) 24`] = `"twenty-three"`;

exports[`numbers match expected output (full range test) 25`] = `"twenty-four"`;

exports[`numbers match expected output (full range test) 26`] = `"twenty-five"`;

exports[`numbers match expected output (full range test) 27`] = `"twenty-six"`;

exports[`numbers match expected output (full range test) 28`] = `"twenty-seven"`;

exports[`numbers match expected output (full range test) 29`] = `"twenty-eight"`;

exports[`numbers match expected output (full range test) 30`] = `"twenty-nine"`;

exports[`numbers match expected output (full range test) 31`] = `"thirty"`;

exports[`numbers match expected output (full range test) 32`] = `"thirty-one"`;

exports[`numbers match expected output (full range test) 33`] = `"thirty-two"`;

exports[`numbers match expected output (full range test) 34`] = `"thirty-three"`;

exports[`numbers match expected output (full range test) 35`] = `"thirty-four"`;

exports[`numbers match expected output (full range test) 36`] = `"thirty-five"`;

exports[`numbers match expected output (full range test) 37`] = `"thirty-six"`;

exports[`numbers match expected output (full range test) 38`] = `"thirty-seven"`;

exports[`numbers match expected output (full range test) 39`] = `"thirty-eight"`;

exports[`numbers match expected output (full range test) 40`] = `"thirty-nine"`;

exports[`numbers match expected output (full range test) 41`] = `"forty"`;

exports[`numbers match expected output (full range test) 42`] = `"forty-one"`;

exports[`numbers match expected output (full range test) 43`] = `"forty-two"`;

exports[`numbers match expected output (full range test) 44`] = `"forty-three"`;

exports[`numbers match expected output (full range test) 45`] = `"forty-four"`;

exports[`numbers match expected output (full range test) 46`] = `"forty-five"`;

exports[`numbers match expected output (full range test) 47`] = `"forty-six"`;

exports[`numbers match expected output (full range test) 48`] = `"forty-seven"`;

exports[`numbers match expected output (full range test) 49`] = `"forty-eight"`;

exports[`numbers match expected output (full range test) 50`] = `"forty-nine"`;

exports[`numbers match expected output (full range test) 51`] = `"fifty"`;

exports[`numbers match expected output (full range test) 52`] = `"fifty-one"`;

exports[`numbers match expected output (full range test) 53`] = `"fifty-two"`;

exports[`numbers match expected output (full range test) 54`] = `"fifty-three"`;

exports[`numbers match expected output (full range test) 55`] = `"fifty-four"`;

exports[`numbers match expected output (full range test) 56`] = `"fifty-five"`;

exports[`numbers match expected output (full range test) 57`] = `"fifty-six"`;

exports[`numbers match expected output (full range test) 58`] = `"fifty-seven"`;

exports[`numbers match expected output (full range test) 59`] = `"fifty-eight"`;

exports[`numbers match expected output (full range test) 60`] = `"fifty-nine"`;

exports[`numbers match expected output (full range test) 61`] = `"sixty"`;

exports[`numbers match expected output (full range test) 62`] = `"sixty-one"`;

exports[`numbers match expected output (full range test) 63`] = `"sixty-two"`;

exports[`numbers match expected output (full range test) 64`] = `"sixty-three"`;

exports[`numbers match expected output (full range test) 65`] = `"sixty-four"`;

exports[`numbers match expected output (full range test) 66`] = `"sixty-five"`;

exports[`numbers match expected output (full range test) 67`] = `"sixty-six"`;

exports[`numbers match expected output (full range test) 68`] = `"sixty-seven"`;

exports[`numbers match expected output (full range test) 69`] = `"sixty-eight"`;

exports[`numbers match expected output (full range test) 70`] = `"sixty-nine"`;

exports[`numbers match expected output (full range test) 71`] = `"seventy"`;

exports[`numbers match expected output (full range test) 72`] = `"seventy-one"`;

exports[`numbers match expected output (full range test) 73`] = `"seventy-two"`;

exports[`numbers match expected output (full range test) 74`] = `"seventy-three"`;

exports[`numbers match expected output (full range test) 75`] = `"seventy-four"`;

exports[`numbers match expected output (full range test) 76`] = `"seventy-five"`;

exports[`numbers match expected output (full range test) 77`] = `"seventy-six"`;

exports[`numbers match expected output (full range test) 78`] = `"seventy-seven"`;

exports[`numbers match expected output (full range test) 79`] = `"seventy-eight"`;

exports[`numbers match expected output (full range test) 80`] = `"seventy-nine"`;

exports[`numbers match expected output (full range test) 81`] = `"eighty"`;

exports[`numbers match expected output (full range test) 82`] = `"eighty-one"`;

exports[`numbers match expected output (full range test) 83`] = `"eighty-two"`;

exports[`numbers match expected output (full range test) 84`] = `"eighty-three"`;

exports[`numbers match expected output (full range test) 85`] = `"eighty-four"`;

exports[`numbers match expected output (full range test) 86`] = `"eighty-five"`;

exports[`numbers match expected output (full range test) 87`] = `"eighty-six"`;

exports[`numbers match expected output (full range test) 88`] = `"eighty-seven"`;

exports[`numbers match expected output (full range test) 89`] = `"eighty-eight"`;

exports[`numbers match expected output (full range test) 90`] = `"eighty-nine"`;

exports[`numbers match expected output (full range test) 91`] = `"ninety"`;

exports[`numbers match expected output (full range test) 92`] = `"ninety-one"`;

exports[`numbers match expected output (full range test) 93`] = `"ninety-two"`;

exports[`numbers match expected output (full range test) 94`] = `"ninety-three"`;

exports[`numbers match expected output (full range test) 95`] = `"ninety-four"`;

exports[`numbers match expected output (full range test) 96`] = `"ninety-five"`;

exports[`numbers match expected output (full range test) 97`] = `"ninety-six"`;

exports[`numbers match expected output (full range test) 98`] = `"ninety-seven"`;

exports[`numbers match expected output (full range test) 99`] = `"ninety-eight"`;

exports[`numbers match expected output (full range test) 100`] = `"ninety-nine"`;

exports[`throws TypeError when a non-string value is passed 1`] = `"_digitsToEnglishTwoDigit only accepts strings"`;

exports[`throws a RangeError when a value with more than two digits is used 1`] = `"_digitsToEnglishTwoDigit accepts at most two characters"`;

exports[`throws a RangeError when a value with more than two digits is used 2`] = `"_digitsToEnglishTwoDigit accepts at most two characters"`;
